src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,547 @@
+		return SC_ERROR_INTERNAL;
+
+	if (!privdata->can[0]) {
+		/* If no CAN is specified in conf, try to decrypt it from enc_can file */
+		r = eoi_decrypt_can(&privdata->enc_can, privdata->can);
+		LOG_TEST_RET(card->ctx, r, "Cannot decrypt CAN");
+	}
+	/* CAN should be 6 chars long */
+	if (strlen(privdata->can) != 6)
+		return SC_ERROR_DECRYPT_FAILED;
+
+	memset(&pace_input, 0, sizeof pace_input);
+	memset(&pace_output, 0, sizeof pace_output);
+
+	pace_input.pin_id = PACE_PIN_ID_CAN;
+	pace_input.pin = (u8 *)privdata->can;
+	pace_input.pin_length = strlen(privdata->can);
+
+	/* EF.CardAccess can only be read from MF */
+	r = sc_select_file(card, sc_get_mf_path(), NULL);
+	LOG_TEST_RET(card->ctx, r, "sc_select_file failed");
+
+	r = perform_pace(card, pace_input, &pace_output, EAC_TR_VERSION_2_02);
+	LOG_TEST_RET(card->ctx, r, "Error verifying CAN");
+
+	return SC_SUCCESS;
+}
+
+static int eoi_get_data(sc_card_t *card, u8 data_id, u8 *buf, size_t len)
+{
+	int r;
+	sc_apdu_t apdu;
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0xCA, 0x01, data_id);
+	apdu.resp = buf;
+	apdu.resplen = len;
+	apdu.le = len;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	return r;
+}
+
+#define ATR_MATCH 1
+
+static int eoi_match_card(sc_card_t* card) {
+	LOG_FUNC_CALLED(card->ctx);
+	if (_sc_match_atr(card, eoi_atrs, &card->type) >= 0) {
+		sc_log(card->ctx, "ATR recognized as Slovenian eID card");
+		LOG_FUNC_RETURN(card->ctx, ATR_MATCH);
+	}
+	LOG_FUNC_RETURN(card->ctx, !ATR_MATCH);
+}
+
+static int eoi_init(sc_card_t* card) {
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	u8 version[6];
+	size_t i, j;
+	scconf_block **found_blocks, *block;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (eoi_get_data(card, 0x16, version, sizeof(version)) != SC_SUCCESS)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_WRONG_CARD);
+
+	if (privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	privdata = sc_mem_secure_alloc(sizeof(struct eoi_privdata));
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+	/* sc_mem_secure_alloc()-ed memory may not be zeroized */
+	memset(privdata, 0, sizeof(struct eoi_privdata));
+	card->drv_data = privdata;
+
+	sprintf(privdata->version, "%X%02X.%02X%02X", version[0], version[1], version[2], version[3]);
+	sc_log(card->ctx, "App version: %s", privdata->version);
+
+	memset(&card->sm_ctx, 0, sizeof card->sm_ctx);
+	card->sm_ctx.ops.open = eoi_sm_open;
+
+	card->max_send_size = SC_MAX_APDU_DATA_SIZE;
+	card->max_recv_size = SC_MAX_APDU_RESP_SIZE;
+
+	for (i = 0; i < sizeof eoi_curves / sizeof * eoi_curves; ++i) {
+		int r = _sc_card_add_ec_alg(card, eoi_curves[i].len, SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ECDSA_HASH_NONE, 0, &eoi_curves[i].oid);
+		LOG_TEST_RET(card->ctx, r, "Add EC alg failed");
+	}
+
+	if (getenv("EOI_CAN"))
+		strncpy(privdata->can, getenv("EOI_CAN"), sizeof(privdata->can) - 1);
```

Comment:
I would prefer to use temporary variable instead of calling the `getenv()` twice. Not sure if compilers are clever enough to optimize this out.
It will also read behind the buffer limits if the env variable will be shorter than the required 6 chars or so.

Reply:
Done. Altough `privdata` is zeroized and thus guaranteed that `privdata->can` ends with `\0` I've replaced `strncpy` with `strlcpy`.